### DISCUSSION TITLE
'Model' object is not iterable

### DISCUSSION THREAD

**Discussion by vittxr** [sqlalchemy/sqlalchemy#10454] [Usage Questions]
Hi! I am developing an endpoint, where I can filter the data using 'q' and 'cols' in the request parameters. (q = query and cols = desired data fields)

to filter the data by query, i use .filter() method and the returned data is: ```<class Model> ``` 
to filter cols, i use with_entities() and the returned data is ```<class sqlalchemy.engine.row>``` 

It's not a big deal, but I think i am doing something that could be done better.

I would like to have the same data type. When I have two returns I have to make conditions to check the data type: 

```python
data = query.all()
data = [dict(row._mapping) if hasattr(row, "_mapping") else row.as_dict() for row in data] # as_dict() is a model method.
```

I can use ```dict(row)``` only if my data type is ``` <class sqlalchemy.engine.row>```

this is my endpoint code: 

```python
@app.route("/")
def test():
    cols = request.args.get("cols")
    q = request.args.get("q")

    query = Model.query
    if q:
        key, value, operator = extract_query_param(q)
        query = sqa_filter(query, Model, key, value, operator)

    if cols:
        cols = [column(col) for col in cols.split(",")]
        query = query.with_entities(*cols)

    data = query.all()
    data = [
        dict(row._mapping) if hasattr(row, "_mapping") else row.as_dict()
        for row in data
    ]

    return {
        "data": data,
    }
```

if  ```cols``` is not none, the following block is executed
```python
if cols:
        cols = [column(col) for col in cols.split(",")]
        query = query.with_entities(*cols)
```
and then my record is parsed to ```<class sqlalchemy.engine.row>```

My specs: 

```txt
Flask==2.2.2
PyMySQL==1.0.2
Flask-SQLAlchemy==3.0.2
SQLAlchemy==2.0.6
python-dotenv==1.0.0
```

in [docs](https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.with_entities), I see that with_entities returns ```Query[Any] ```.

I was expecting do something like: 
```python
data = [dict(row) for row in data]
```

In place of dict(), I tried: 
- ```._mapping()```: doesnt work in <Model> 
- ```.__dict__```: doesnt work in <class sqlalchemy.engine.row>

code: [sqlalchemy.zip](https://github.com/sqlalchemy/sqlalchemy/files/12871468/sqlalchemy.zip)

**Comment by zzzeek** [comment]
the code you have is not super efficient and could be improved, but seems like it would work as written.  I'm not following what the issue is? What does "doesnt work" mean?   stack trace?

**Reply by vittxr** [reply]
There is no error because I use if conditions to check the returned datatype of the SQLAlchemy method. However, as you said, it could be improved. I think I didn't title this question very well, but the "'Model' object is not iterable" error occurs when I execute the `with_entities()` method. In this case, I need to use `dict(row._mapping)` to convert the Row object to a dictionary. Otherwise, when I don't use `with_entities()` (only the `filters()` method), the datatype returned is the model that I created, and it doesn't have a `._mapping` attribute. In this case, I use a custom `as_dict()` method that I implemented in the SQLAlchemy model.

I was wondering if it's possible to have the same datatype after executing these methods so that I don't need to use an if condition

This is the reason that I thought my SQLAlchemy code is wrong

**Reply by zzzeek** [reply]
So there's a lot of different angles on this, including that you could update it to use [2.0 style queries](https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.column_descriptions) to use newer patterns.

however, you still have to know if you are going to be parsing out individual columns or ORM entities based on how your function receives input.   it's also not clear if when someone sends your function "cols", is that SQL columns or also can that include ORM entities.

if "cols" is always SQL expressions only, and "q" is always a single ORM entity, you could just determine up front based on that:

```py
@app.route("/")
def test():
    cols = request.args.get("cols")
    q = request.args.get("q")

    query = Model.query
    if q:
        key, value, operator = extract_query_param(q)
        query = sqa_filter(query, Model, key, value, operator)

    if cols:
        cols = [column(col) for col in cols.split(",")]
        query = query.with_entities(*cols).only_return_tuples()

        data = [
            dict(row._mapping) for row in query
        ]
    else:
        data = [
            obj.as_dict() dfor obj in query
        ]

    return {
        "data": data,
    }
```


only_return_tuples() shouldn't be needed above but it ensures you get tuples for the "cols" case.

Whether you get back scalars or tuples is easier to reason about if you use 2.0 style queries.

**Comment by CaselIT** [comment]
Hi,

Depending on how your asdict is implemented, you could use the option load_only to just load the columns you want https://docs.sqlalchemy.org/en/20/orm/queryguide/columns.html#sqlalchemy.orm.load_only

**Comment by vittxr** [comment]
Hi, guys. Thanks for the replies, but I solved it using:
```select()```
and
```with_only_columns()```
from version 2.0

I didn't use ```scalars()``` because I want to retrieve all the data, not only the first column, and load_only seems to be used with the session.query(model) syntax style.

My route after all is:
```python
@app.route("/")
def test():
    cols = request.args.get("cols")
    q = request.args.get("q")
    query = db.select(ModelExample)

    if q:
        key, value, operator = extract_query_param(q)
        query = sqa_filter(query, ModelExample, key, value, operator)

    if cols:
        query = query.with_only_columns(
            *[getattr(ModelExample, col) for col in cols.split(",")]
        )

    data = db.session.execute(query.offset(0).limit(10)).all()
    data = [row._asdict() for row in data]  # row_asdict() = dict(row._mapping)
    return jsonify({"data": data})
```
I think it's in 2.x style.

**Reply by CaselIT** [reply]
>  load_only seems to be used with the session.query(model) syntax style.

you can use load_only and any other loader also when using select.

But if the above works then it's fine too. There are more ways to pet a cat :)

**Reply by vittxr** [reply]
Oh, true. I can use ```query = query.options(
            load_only(*[getattr(ModelExample, col) for col in cols.split(",")])
        )```, but convert it to dict then is slightly different
